docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,268 @@
+
+### Risks and Mitigations
+
+Example risks:
+1. A bad snapshot is taken, and we will mitigate this risk by having trying to prevent it before it happens by checking if things are ready and draining everything before taking the snapshot. Rolling back and determining if a bad snapshot is bad is out of scope for this proposal currently and will be for the operator to watch, so here we will simply try to prevent it as best as we can.
+1. A bad security patch or update might have been applied to a user’s node that they don’t want to be applied to future nodes. To mitigate this risk, we will make it easy for users to turn this feature off, and if they fix it on their original node the snapshot will be taken of that node instead.
+1. Deleting previous snapshots might break things since Compute Gallery Image Definition Version instances may depend directly on those snapshots still being there. A way to mitigate this is to implement the deletion with the optional field for rolling update and if the prototyping is non-rolling slowly make all the images of the previously instantiated azuremachinepoolmachines match the current image, and when the previous image isn't being used anymore delete that version and the snapshot associated with it.
+
+The following limits exist for Azure Compute Galleries:
+1. 100 galleries, per subscription, per region
+1. 1,000 image definitions, per subscription, per region
+1. 10,000 image versions, per subscription, per region
+1. 100 replicas per image version however 50 replicas should be sufficient for most use cases
+1. Any disk attached to the image must be less than or equal to 1 TB in size
+1. Resource move isn't supported for Azure compute gallery resources
+
+Link to page with Azure Compute Gallery limits: https://learn.microsoft.com/en-us/azure/virtual-machines/azure-compute-gallery
+
+Thus the feature will not work for OS images with disks attached greater than 1 TB in size. These limits should be kept in mind by the operator since this feature requires the use of an Azure Compute Gallery. The Azure Compute Gallery itself costs no money. For each image, it costs roughly $0.10 - $0.13 for a CAPI image per region and $0.1 - $0.3 for a plain Flatcar image per region with the Standard HDD LRS storage account type every day. Scaling across other regions will also accrue network egress charges. Thus, the feature may cost the operator more money if they don't need faster horizontal scaling.
```

Comment:
Do `$0.1` and `$0.3` mean 10 cents and 30 cents, or is this a typo and it should be `$0.01` and `$0.03`? Either way, two digits for cents seems clearest.

Reply:
It was because of how an email was sent to me about these prices with this formatting, fixed!